Sources/Algorithms/EndsWith.swift

```diff
@@ -0,0 +1,95 @@
+  ///
+  /// This example tests whether one countable range ends with the elements
+  /// of another countable range.
+  ///
+  ///     let a = 8...10
+  ///     let b = 1...10
+  ///
+  ///     print(b.ends(with: a))
+  ///     // Prints "true"
+  ///
+  /// Passing a collection with no elements or an empty collection as
+  /// `possibleSuffix` always results in `true`.
+  ///
+  ///     print(b.ends(with: []))
+  ///     // Prints "true"
+  ///
+  /// - Parameter possibleSuffix: A collection to compare to this collection.
+  /// - Returns: `true` if the initial elements of the collection are the same as
+  ///   the elements of `possibleSuffix`; otherwise, `false`. If
+  ///   `possibleSuffix` has no elements, the return value is `true`.
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   collection and the length of `possibleSuffix`.
+  // Adapted from https://github.com/apple/swift/blob/d6f9401/stdlib/public/core/SequenceAlgorithms.swift#L281-L286
+  @inlinable
+  public func ends<PossibleSuffix: BidirectionalCollection>(
+    with possibleSuffix: PossibleSuffix
+  ) -> Bool where PossibleSuffix.Element == Element {
+    return self.ends(with: possibleSuffix, by: ==)
+  }
+
+  /// Returns a Boolean value indicating whether the final elements of the
+  /// collection are equivalent to the elements in another collection, using
+  /// the given predicate as the equivalence test.
+  ///
+  /// The predicate must be a *equivalence relation* over the elements. That
+  /// is, for any elements `a`, `b`, and `c`, the following conditions must
+  /// hold:
+  ///
+  /// - `areEquivalent(a, a)` is always `true`. (Reflexivity)
+  /// - `areEquivalent(a, b)` implies `areEquivalent(b, a)`. (Symmetry)
+  /// - If `areEquivalent(a, b)` and `areEquivalent(b, c)` are both `true`, then
+  ///   `areEquivalent(a, c)` is also `true`. (Transitivity)
+  ///
+  /// - Parameters:
+  ///   - possibleSuffix: A collection to compare to this collection.
+  ///   - areEquivalent: A predicate that returns `true` if its two arguments
+  ///     are equivalent; otherwise, `false`.
+  /// - Returns: `true` if the initial elements of the collection are equivalent
+  ///   to the elements of `possibleSuffix`; otherwise, `false`. If
+  ///   `possibleSuffix` has no elements, the return value is `true`.
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   collection and the length of `possibleSuffix`.
+  // Adapted from https://github.com/apple/swift/blob/d6f9401/stdlib/public/core/SequenceAlgorithms.swift#L235-L252
+  @inlinable
+  public func ends<PossibleSuffix: BidirectionalCollection>(
+    with possibleSuffix: PossibleSuffix,
+    by areEquivalent: (Element, PossibleSuffix.Element) throws -> Bool
+  ) rethrows -> Bool {
+    var possibleSuffixIterator = possibleSuffix.reversed().makeIterator()
+    for e0 in self.reversed() {
+      if let e1 = possibleSuffixIterator.next() {
```

Comment:
Is there an efficient way to implement this without needing backwards iteration of the `possibleSuffix`?

The only way I can think of is basically a [substring search](https://en.wikipedia.org/wiki/String-searching_algorithm), and checking if it ends up match against the very end.

If such an algorithm existed, it would allow us to loosen the requirement of `possibleSuffix` from `BidirectionalCollection` down to just `Sequence`.

Reply:
If we loosen the requirement to just `Collection` (so that we have a known `count`), then we can probably iterate from the middle (starting at `self.index(self.endIndex, offsetBy: possibleSuffix.count`), and matching until the end. I'm not sure if this is better though, because we need to do this (potentially-non-constant-time) index math.

Reply:
Right … I don't imagine that there are many cases where you have a bidirectional collection but you can't make the suffix you're searching for bidirectional.

Reply:
I agree.

For completeness, I'll mention an alternative implementation I toyed with, which has different constraints:

1. `self` needs to be `RandomAccessCollection` instead of `Bidirectional` (and it does some indexing offsetting, which can be slow if that's not `O(1)`).
2. The `possiblePrefix` only needs to be a `Collection` instead of a `BidirectionalCollection`

```swift
extension RandomAccessCollection where Element: Equatable {
  @inlinable
  public func ends<PossibleSuffix: Collection>(
    with possibleSuffix: PossibleSuffix,
    by areEquivalent: (Element, PossibleSuffix.Element) throws -> Bool
  ) rethrows -> Bool {
    guard let startOfSuffix = self.index(self.endIndex, offsetBy: -possibleSuffix.count, limitedBy: self.startIndex) else {
      return false
    }

    return try self[startOfSuffix...].elementsEqual(possibleSuffix, by: areEquivalent)
  }
}
```

Reply:
Just a small addition here is that in this case `possibleSuffix` probably will be better if constrained to `RandomAccessCollection` as well, otherwise `-possibleSuffix.count` is not guaranteed to be O(1).